src/XMonad/Operations.hs

```diff
@@ -565,7 +565,10 @@ floatLocation w =
           ws <- gets windowset
           wa <- io $ getWindowAttributes d w
           let bw = (fromIntegral . wa_border_width) wa
-          sc <- fromMaybe (W.current ws) <$> pointScreen (fi $ wa_x wa) (fi $ wa_y wa)
+          managed <- isClient w
+          sc <- if managed
+                then fromMaybe (W.current ws) <$> pointScreen (fi $ wa_x wa) (fi $ wa_y wa)
+                else return (W.current ws)
 
           let sr = screenRect . W.screenDetail $ sc
               rr = W.RationalRect ((fi (wa_x wa) - fi (rect_x sr)) % fi (rect_width sr))
```

Comment:
The above now computes negative position coordinates on all screens other than the one with `(wa_x wa, wa_y wa)` (usually 0, 0). This is then fixed by `adjust` in `manage`, but other uses of floatLocation (doFloat, …) do not make those adjustments. This means that a `doFloat` in manage hook will result in the window being placed at 0, 0 even though it belongs to a workspace on another screen.

To fix this, we must not use the screen rect if we didn't use pointScreen to obtain the correct screen rect in the first place. Perhaps just place the float at the center. I'll try making this change and will continue testing with that.

Also, we might consider moving the `adjust` function here.

Reply:
https://github.com/liskin/xmonad/commit/207173d83ad107f8e610ceff300d0a1618fae13e

Reply:
Good catch. I was using doCenterFloat instead of doFloat, so that issue didn't occur for me. 
I don't think adjust should go straight up in floatLocation however, as there are still times where you want windows to be able to go beyond screen boundaries (i.e. when dragging a window from screen to screen with the mouse.)

Reply:
Oh, right, dragging does indeed behave erratically now. I'll try something else later today.

Reply:
How about this: https://github.com/liskin/xmonad/commit/9bcb0c3219108bc33ccdfb20edf9a181a0208dc5 ?

I moved `adjust` back to `manage` and changed `floatLocation` to only adjust when the window is unmanaged and its x,y coordinates are outside of current screen. This should fix the flickering and should also stop forcing float centering of windows that happen to be placed in the correct screen. (And I added explicit centering to my managehook as I want that behaviour anyway.)

Reply:
Another fix for flickering of dragging windows, this time when dragging them outside screen: 48b0caa613922124ffcadb2b112fcd1d6018061d